engine-tests/src/test/java/org/terasology/engine/rendering/world/ChunkMeshWorkerTest.java

```diff
@@ -23,107 +29,158 @@
     }
 
     @Test
-    void testChunksDirtiedBetweenGenerationAndUploadAreUploadedAnyway() {
-        // maybe redundant with testChunkIsRegeneratedIfDirty?
-        // I guess the assertion is that the upload function did a thing
-        // instead of skipping a thing.
+    void testChunkIsNotProcessedTwice() {
+        var chunk1 = newDirtyChunk(position0);
+
+        verifier.then(() -> {
+                    worker.add(chunk1);
+                    worker.add(chunk1);  // added twice
+                    worker.update();
+                })
+                .expectNextCount(1).as("expect only one result")
+                .then(() -> {
+                    // adding it again and doing another update should still not change
+                    worker.add(chunk1);
+                    worker.update();
+                })
+                .verifyTimeout(Duration.ofSeconds(5));
     }
 
     @Test
-    void testChunkCanBeRemovedBeforeMeshGeneration() {
-        var chunk = new DummyChunk(position0);
-        worker.add(chunk);
-        worker.remove(chunk);
-        // drain
-        fail("TODO: assert no work happened on chunk");
+    void testChunkIsRegeneratedIfDirty() {
+        var chunk1 = newDirtyChunk(position0);
+
+        verifier.then(() -> {
+                    worker.add(chunk1);
+                    worker.update();
+                })
+                .expectNext(chunk1).as("initial generation")
+                .then(() -> {
+                    chunk1.setDirty(true);
+                    worker.update();
+                })
+                .expectNext(chunk1).as("regenerating after dirty")
+                .verifyTimeout(Duration.ofSeconds(5));
     }
 
     @Test
-    void testDoubleRemoveIsNoProblem() {
-        var chunk = new DummyChunk(position0);
-        worker.add(chunk);
-        worker.remove(chunk);
-        worker.remove(chunk);
-        // drain
+    void testChunkCanBeRemovedBeforeMeshGeneration() {
+        var chunk = newDirtyChunk(position0);
+        verifier.then(() -> {
+                    worker.add(chunk);
+                    worker.remove(chunk);
+                    worker.update();
+                })
+                // chunk was removed, no events expected
+                .verifyTimeout(Duration.ofSeconds(5));
     }
 
     @Test
-    void testChunkCanBeRemovedBeforeUpload() {
-        var chunk = new DummyChunk(position0);
-        worker.add(chunk);
-        // tick so generation is finished, but not upload
-        worker.remove(chunk);
-        // drain
-        fail("assert upload did not happen for chunk");
+    void testDoubleRemoveIsNoProblem() {
+        var chunk = newDirtyChunk(position0);
+        verifier.then(() -> {
+                    worker.add(chunk);
+                    worker.remove(chunk);
+                    worker.update();
+                })
+                .then(() -> {
+                    worker.remove(chunk);  // second time calling remove on the same chunk
```

Comment:
do we consider the following a different test case?
```
worker.add(chunk);
worker.remove(chunk);
worker.remove(chunk);
worker.update();
```

Reply:
I don't see one, no. Are you concerned we need one?